drivers/flash/flash_stm32l5x.c

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright (c) 2021 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_DOMAIN flash_stm32l5
+#define LOG_LEVEL CONFIG_FLASH_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_DOMAIN);
+
+#include <kernel.h>
+#include <device.h>
+#include <string.h>
+#include <drivers/flash.h>
+#include <init.h>
+#include <soc.h>
+#include <stm32l5xx_ll_icache.h>
+#include <stm32_ll_system.h>
+
+#include "flash_stm32.h"
+
+#define STM32L5_SERIES_MAX_FLASH	512
+#define BANK2_OFFSET	(KB(STM32L5_SERIES_MAX_FLASH) / 2)
+
+#define ICACHE_DISABLE_TIMEOUT_VALUE           1U   /* 1ms */
+#define ICACHE_INVALIDATE_TIMEOUT_VALUE        1U   /* 1ms */
+
+static int stm32l5x_icache_disable(void)
+{
+	int status = 0;
+	uint32_t tickstart;
+
+	LOG_DBG("I-cache Disable");
+	/* Clear BSYENDF flag first and then disable the instruction cache that starts
+	 * a cache invalidation procedure
+	 */
+	CLEAR_BIT(ICACHE->FCR, ICACHE_FCR_CBSYENDF);
+
+	CLEAR_BIT(ICACHE->CR, ICACHE_CR_EN);
+
+	/* Get tick */
+	tickstart = HAL_GetTick();
+
+	/* Wait for instruction cache being disabled */
+	while (READ_BIT(ICACHE->CR, ICACHE_CR_EN) != 0U) {
+		if ((HAL_GetTick() - tickstart) > ICACHE_DISABLE_TIMEOUT_VALUE) {
+			/* New check to avoid false timeout detection in case
+			 * of preemption.
+			 */
+			if (READ_BIT(ICACHE->CR, ICACHE_CR_EN) != 0U) {
+				status = -ETIMEDOUT;
+				break;
+			}
+		}
+	}
+
+	return status;
+}
+
+static void stm32l5x_icache_enable(void)
+{
+	LOG_DBG("I-cache Enable");
+	SET_BIT(ICACHE->CR, ICACHE_CR_EN);
+}
+
+static int icache_wait_for_invalidate_complete(void)
+{
+	int status = -EIO;
+	uint32_t tickstart;
+
+	/* Check if ongoing invalidation operation */
+	if (READ_BIT(ICACHE->SR, ICACHE_SR_BUSYF) != 0U) {
+		/* Get tick */
+		tickstart = HAL_GetTick();
+
+		/* Wait for end of cache invalidation */
+		while (READ_BIT(ICACHE->SR, ICACHE_SR_BSYENDF) == 0U) {
+			if ((HAL_GetTick() - tickstart)
+					> ICACHE_INVALIDATE_TIMEOUT_VALUE) {
+				/* New check to avoid false timeout detection
+				 * in case of preemption
+				 */
+				if (READ_BIT(ICACHE->SR, ICACHE_SR_BSYENDF)
+									== 0U) {
+					status = -ETIMEDOUT;
+					break;
+				}
+			}
+		}
+	}
+
+	/* Clear any pending flags */
+	if (LL_ICACHE_IsActiveFlag_BSYEND()) {
```

Comment:
This is the 3rd time in this small function that we check the same flag.
I understand it is done twice, like in HAL, but 3 times is too much.
Please reorganize the function to remove at least one occurrence.

Reply:
I understand it is useful in you implementation, but I am almost sure it is possible to reorganize the function to simplify this function and remove at least one check. 
For example (and this is only an example), **return -ETIMEDOUT;** instead of **break;**  could guarantee that flasg is set